src/include/migraphx/instruction_ref.hpp

```diff
@@ -32,7 +32,47 @@ namespace migraphx {
 inline namespace MIGRAPHX_INLINE_NS {
 
 struct instruction;
+#if defined(_WIN32) && !defined(NDEBUG)
+struct instruction_ref : public std::list<instruction>::iterator
+{
+    using instruction_iter = std::list<instruction>::iterator;
+    using instruction_const_iter = std::list<instruction>::const_iterator;
+
+    instruction_ref() = default;
+    instruction_ref(const instruction_iter& other)
+    {
+        _Ptr = other._Ptr;
+        _Adopt(other._Getcont());
+    }
+
+    friend bool operator==(const instruction_ref& x, const instruction_ref& y);
+    friend bool operator==(const instruction_ref& x, const instruction_iter& y);
+    friend bool operator==(const instruction_ref& x, const instruction_const_iter& y);
```

Comment:
Is this overload needed as well?

Reply:
Yes, a comparison between an instance of `instruction_ref` and an .end( ) of a list is performed somewhere in the code, hence the operator overload was necessary.

Reply:
Yea, thats why it should be `friend bool operator==(const std::list<instruction>::iterator& x, const std::list<instruction>::iterator& y)` so it will work with both `std::list<instruction>::iterator` and `instruction_ref`.

Reply:
We arent overriding whats in the standard library.  A friend function is not visible to the outer scope namespace, it can only be called through ADL, which means you need to be using an `instruction_ref` argument.

Reply:
Even though we don't overload what is in STL, there is an ambiguity when it comes to choosing the comparison operator, as an already fitting one is found in the STL:
```
C:/develop/AMDMIGraphX/src/module.cpp:512:12: error: use of overloaded operator '!=' is ambiguous (with operand types 'std::_List_iterator<std::_List_val<std::_List_simple_types<migraphx::instruction>>>' and 'instruction_ref')
    if(ins != this->end())
       ~~~ ^  ~~~~~~~~~~~
C:/develop/AMDMIGraphX/src/include\migraphx/instruction_ref.hpp:63:17: note: candidate function
    friend bool operator!=(const instruction_iter& x, const instruction_iter& y)
                ^
C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.35.32215\include\list:202:21: note: candidate function
    _NODISCARD bool operator!=(const _List_const_iterator& _Right) const noexcept {
                    ^
C:/develop/AMDMIGraphX/src/include\migraphx/instruction_ref.hpp:58:17: note: candidate function
    friend bool operator!=(const instruction_ref& x, const instruction_ref& y)
                ^
```

Reply:
That is because you have two overloads that match at line 58 and line 63. There should only be one `operator==` defined:

```cpp
friend bool operator==(const std::list<instruction>::iterator& x, const std::list<instruction>::iterator& y)
{
    ...
}
```

If that doesnt resolve the ambiguity then we could try a template:

```cpp
template<class T, class U, MIGRAPHX_REQUIRES(std::is_same<T, instruction_ref>{} or std::is_same<U, instruction_ref>{})>
friend bool operator==(const T& x, const U& y)
{
    ...
}
```

Otherwise you are going to need 5 overloads, so it will be simpler if you can just define one.

Reply:
It's done. Trying the first option with `std::list<instruction>::iterator`s was causing the function body to never be invoked, however the second approach with conditional enablement did the job.